### PR TITLE
#5377: sync revoked packages and show unavailable on mod screen

### LINKED ISSUE #5377: Fix handling of deleted mods and mods where access has been revoked
Context
---
- https://pixiebrix.zendesk.com/agent/tickets/254
- https://pixiebrix.slack.com/archives/C0436P48QHY/p1679430537174119

Additional Problems
---
- Extension Console crashes in some cases (related?): https://pixiebrix.slack.com/archives/C0436P48QHY/p1679504564622899

Discussion
---
- The issue occurs because we switched to `fetchNewPackages` which does not remove packages that no longer exist/that the user no longer has access to
https://github.com/pixiebrix/pixiebrix-extension/blob/a982af8c48eb3c80decd066a7d2f5886b9761fe2/src/baseRegistry.ts#L63-L63
- Originally fetchNewPackages was supposed to be a performance optimization, but it leads to correctness problems on what user should have access to
- Removing the use of fetchNewPackages will revive the old bug we had in Page Editor where deleted mods will shows as "Loading..." permanently, and there's no way to remove the mod because it won't show up in the Mods screen

Related Issues
---
- https://github.com/pixiebrix/pixiebrix-extension/issues/4215

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #5377 
- Removes `fetchNewPackages` from the extension. Extension always uses syncPackages now to ensure deleted/revoked packages are removed locally
- Adds an `UnavailableRecipe` type to represent mods the user no longer has access to on the mods screen
- Adds ErrorBoundaries around the Grid and List items to add resilience on per-card/item rendering

## Review Tips

- Review the switch from `fetchNewPackages` to `syncPackages`
- Review the changes to `useInstallables.ts`
- Review changes to `installableUtils.ts`
- Review changes to other files

## Demo

- https://www.loom.com/share/48e54db087b54d168b085b63345f1b1c

## Future Work

- Handle the stuck "Loading..." state in the Page Editor for unavailable mods

## Checklist

- [x] Add tests
- [x] Designate a primary reviewer: @BLoe

### CODE REVIEW

File: src/registry/localRegistry.ts [hunk 1]
```diff
@@ -126,22 +124,6 @@ export async function clear(): Promise<void> {
   await db.clear(BRICK_STORE);
 }
 
-/**
- * Return the timestamp of the most recent package in the local DB.
- */
-async function latestTimestamp(): Promise<Date | null> {
```
Comment (twschiller) [inline src/registry/localRegistry.ts:132]: Drop unused code

File: src/registry/localRegistry.ts [hunk 2]
```diff
@@ -177,42 +159,6 @@ export function parsePackage(
...
 }
 
-/**
- * Fetch all new packages from the registry and put them in the local database.
- *
- * Memoized to avoid multiple network requests across tabs.
- *
- * @returns true if the registry was updated, false otherwise
- */
-export const fetchNewPackages = memoizeUntilSettled(async () => {
```
Comment (twschiller) [inline src/registry/localRegistry.ts:187]: Drop unused code

File: src/store/extensionsSlice.ts [hunk 3]
```diff
@@ -70,6 +70,14 @@ const extensionsSlice = createSlice({
       state.extensions = [];
     },
 
+    // Helper reducers method to directly update extensions in tests
```
Comment (twschiller) [inline src/store/extensionsSlice.ts:73]: Our renderHook helper doesn't re-expose a way to provide initial state to the store

File: src/registry/localRegistry.test.ts [hunk 4]
```diff
@@ -39,25 +38,11 @@ describe("localRegistry", () => {
     getApiClientMock.mockResolvedValue({
       get: jest.fn().mockResolvedValue({ data: [recipeDefinitionFactory()] }),
     });
-    await fetchNewPackages();
+    await syncPackages();
     const recipes = await getByKinds(["recipe"]);
     expect(recipes).toHaveLength(1);
   });
 
-  it("should fetch new packages", async () => {
```
Comment (twschiller) [inline src/registry/localRegistry.test.ts:47]: Drop test that was testing dropped code

File: src/options/pages/blueprints/utils/installableUtils.ts
Comment (twschiller) [inline src/options/pages/blueprints/utils/installableUtils.ts:1]: Rewrote the function signatures in this file to use `function` for readability

File: src/options/pages/blueprints/useInstallables.test.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    useAllRecipesMock.mockReset();
+    useAllRecipesMock.mockReturnValue({ data: undefined } as any);
+  });
+
+  it("handles empty state", async () => {
+    const wrapper = renderHook(() => useInstallables());
+
+    await wrapper.act(async () => {
+      // NOP -- wait for async state to resolve
+    });
```
Comment (BLoe) [inline src/options/pages/blueprints/useInstallables.test.ts:66]: Was this required to get all these tests working? Did you try `await waitForEffect()` first? Maybe we should add this as a helper on the custom `render` output as well? Just seems like an awkward syntax to remember to use in every test.

File: src/options/pages/blueprints/utils/installableUtils.test.ts [hunk 6]
```diff
@@ -136,3 +141,27 @@ describe("getSharingType", () => {
...
+  it("returns false for a recipe definition", () => {
+    const installable = recipeDefinitionFactory();
+    expect(isUnavailableRecipe(installable)).toBe(false);
+  });
+
+  it("returns false for an extension", () => {
+    const installable = extensionFactory() as ResolvedExtension;
+    expect(isUnavailableRecipe(installable)).toBe(false);
+  });
+});
```
Comment (BLoe) [inline src/options/pages/blueprints/utils/installableUtils.test.ts:167]: Should we add a `true` result test case here?

File: src/store/extensionsSlice.ts (hunk 3 above)
Comment (BLoe) [reply src/store/extensionsSlice.ts:73]: Are you not able to use the existing `installCloudExtension` action for this?

**Review by BLoe — APPROVED** [review]
Not totally sure why we need that new "UNSAFE" redux action, but LGTM otherwise 👍

File: src/store/extensionsSlice.ts (hunk 3 above)
Comment (twschiller) [reply src/store/extensionsSlice.ts:73]: > Are you not able to use the existing installCloudExtension action for this?

Good thought, but technically not due to the typics - CloudExtensions don't support a `_recipe`

Comment (BLoe) [reply src/store/extensionsSlice.ts]: It doesn't look like you're setting any specific data on the extensions where this is used, so what about using `recipeDefinitionFactory()` and just installing a recipe in the tests, if you need recipe extensions?

Comment (BLoe) [reply src/store/extensionsSlice.ts]: Ah, the one with the matching id for the extension would probably be tricky... Oh well... 🤷‍♂️

**Pull request merged**